core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1799 @@
+
+      val isWitScript = WitnessScriptPubKey.isWitnessScriptPubKey(script.asm)
+      val redeemScript =
+        inputMaps(index).getRecords[InputPSBTRecord.RedeemScript](
+          PSBTInputKeyId.RedeemScriptKeyId)
+      val hasWitScript = redeemScript.size == 1 && WitnessScriptPubKey
+        .isWitnessScriptPubKey(redeemScript.head.redeemScript.asm)
+
+      if (!isWitScript && hasWitScript) {
+        previousElements.filterNot(
+          _.key.head == PSBTInputKeyId.WitnessScriptKeyId.byte) :+ InputPSBTRecord
+          .WitnessScript(script.asInstanceOf[RawScriptPubKey])
+      } else {
+        previousElements.filterNot(
+          _.key.head == PSBTInputKeyId.RedeemScriptKeyId.byte) :+ InputPSBTRecord
+          .RedeemScript(script)
+      }
+    }
+
+    val newMap = InputPSBTMap(elements).compressMap(transaction.inputs(index))
+    val newInputMaps = inputMaps.patch(index, Seq(newMap), 1)
+
+    PSBT(globalMap, newInputMaps, outputMaps)
+  }
+
+  /**
+    * Adds script to the indexed OutputPSBTMap to either the RedeemScript
+    * or WitnessScript field depending on the script and available information in the PSBT
+    * @param script ScriptPubKey to add to PSBT
+    * @param index index of the OutputPSBTMap to add script to
+    * @return PSBT with added script
+    */
+  def addScriptToOutput(script: ScriptPubKey, index: Int): PSBT = {
+    require(
+      index < outputMaps.size,
+      s"index must be less than the number of output maps present in the psbt, $index >= ${outputMaps.size}")
+    require(!isFinalized, "Cannot update a PSBT that is finalized")
+
+    val elements = {
+      val previousElements = outputMaps(index).elements
+
+      val isWitScript = WitnessScriptPubKey.isWitnessScriptPubKey(script.asm)
+      val redeemScript =
+        outputMaps(index).getRecords[OutputPSBTRecord.RedeemScript](
+          PSBTOutputKeyId.RedeemScriptKeyId)
+      val hasWitScript = redeemScript.size == 1 && WitnessScriptPubKey
+        .isWitnessScriptPubKey(redeemScript.head.redeemScript.asm)
+
+      if (!isWitScript && hasWitScript) {
+        previousElements.filterNot(
+          _.key.head == PSBTOutputKeyId.WitnessScriptKeyId.byte) :+ OutputPSBTRecord
+          .WitnessScript(script)
+      } else {
+        previousElements.filterNot(
+          _.key.head == PSBTOutputKeyId.RedeemScriptKeyId.byte) :+ OutputPSBTRecord
+          .RedeemScript(script)
+      }
+    }
+
+    val newMap = OutputPSBTMap(elements)
+    val newOutputMaps = outputMaps.patch(index, Seq(newMap), 1)
+
+    PSBT(globalMap, inputMaps, newOutputMaps)
+  }
+
+  /**
+    * Adds the BIP32Path to the indexed InputPSBTMap to the BIP32DerivationPath field
+    * @param extKey ExtKey to derive key from
+    * @param path path of key to add to PSBT
+    * @param index index of the InputPSBTMap to add the BIP32Path to
+    * @return PSBT with added BIP32Path
+    */
+  def addKeyPathToInput(extKey: ExtKey, path: BIP32Path, index: Int): PSBT = {
```

Comment:
We don't current do any `require`s, that seems like it may be a good idea however, it is not specified in the BIP that an `xpub` must be defined to add a `BIP32Path`.